include/libpmemobj++/experimental/persistent_concurrent_hash_map.hpp

```diff
@@ -0,0 +1,1716 @@
+        friend class interface5::persistent_concurrent_hash_map;
+#else
+    public: // workaround
+#endif
+        //! concurrent_hash_map over which we are iterating.
+        const Container *my_map;
+
+        //! Index in hash table for current item
+        size_t my_index;
+
+        //! Pointer to bucket
+        const bucket *my_bucket;
+
+        //! Pointer to node that has current item
+        node *my_node;
+
+        persistent_hash_map_iterator( const Container &map, size_t index, const bucket *b, node_base *n );
+
+    public:
+        //! Construct undefined iterator
+        persistent_hash_map_iterator() : my_map(), my_index(), my_bucket(), my_node() {}
+        persistent_hash_map_iterator( const persistent_hash_map_iterator<Container, typename Container::value_type> &other ) :
+            my_map( other.my_map ),
+            my_index( other.my_index ),
+            my_bucket( other.my_bucket ),
+            my_node( other.my_node )
+        {}
+        Value& operator*() const {
+            __TBB_ASSERT( persistent_hash_map_base::is_valid( my_node ), "iterator uninitialized or at end of container?" );
+            return my_node->item;
+        }
+        Value* operator->() const { return &operator*(); }
+        persistent_hash_map_iterator& operator++();
+
+        //! Post increment
+        persistent_hash_map_iterator operator++( int ) {
+            persistent_hash_map_iterator old( *this );
+            operator++();
+            return old;
+        }
+    };
+
+    template<typename Container, typename Value>
+    persistent_hash_map_iterator<Container, Value>::persistent_hash_map_iterator( const Container &map, size_t index, const bucket *b, node_base *n ) :
+        my_map( &map ),
+        my_index( index ),
+        my_bucket( b ),
+        my_node( static_cast<node*>(n) )
+    {
+        if ( b && !persistent_hash_map_base::is_valid( n ) )
+            advance_to_next_bucket();
+    }
+
+    template<typename Container, typename Value>
+    persistent_hash_map_iterator<Container, Value>& persistent_hash_map_iterator<Container, Value>::operator++() {
+        my_node = static_cast<node*>(my_node->next.get( (my_map->my_pool_uuid) ) );
+        if ( !my_node ) advance_to_next_bucket();
+        return *this;
+    }
+
+    template<typename Container, typename T, typename U>
+    bool operator==( const persistent_hash_map_iterator<Container, T>& i, const persistent_hash_map_iterator<Container, U>& j ) {
+        return i.my_node == j.my_node && i.my_map == j.my_map;
+    }
+
+    template<typename Container, typename T, typename U>
+    bool operator!=( const persistent_hash_map_iterator<Container, T>& i, const persistent_hash_map_iterator<Container, U>& j ) {
+        return i.my_node != j.my_node || i.my_map != j.my_map;
+    }
+
+    //! Range class used with persistent_concurrent_hash_map
+    /** @ingroup containers */
+    template<typename Iterator>
+    class hash_map_range {
```

Comment:
we should probably unify the interfaces with this functionality. in the array it's `slice`